src/openassetio-core/include/openassetio/BatchElementError.hpp

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 The Foundry Visionmongers Ltd
+#pragma once
+
+#include <utility>
+
+#include <openassetio/errors.h>
+#include <openassetio/export.h>
+#include <openassetio/typedefs.hpp>
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+/**
+ * Structure representing per-element batch operation errors.
+ *
+ * Many OpenAssetIO API functions take multiple inputs, i.e. a batch of
+ * elements, in order to allow the @ref manager backend to optimise bulk
+ * queries. The results of such queries are usually returned
+ * element-by-element via a callback.
+ *
+ * It is possible for the whole batch to fail due to some catastrophic
+ * error, in which case a standard exception workflow is expected.
+ *
+ * However, it is also possible for a subset of elements in the batch to
+ * fail, whilst the remainder succeed. An exception workflow doesn't
+ * work so well here, and so every success callback is paired with an
+ * error callback, allowing per-element errors to be communicated back
+ * to the original caller (i.e. the @ref host application).
+ *
+ * The information for these per-element errors is bundled in instances
+ * of this simple BatchElementError structure for passing to error
+ * callbacks.
+ *
+ * This structure provides an error code, for control flow, and an
+ * error message, for more (human-readable) detail.
+ */
+class BatchElementError final {
+ public:
+  /// Possible classes of error.
+  enum class ErrorCode {
+    /// Fallback for uncommon errors.
+    kUnknown = OPENASSETIO_BatchErrorCode_kUnknown
+  };
+
+  /// Error code indicating the class of error.
+  const ErrorCode code;
+  /// Human-readable error message.
+  const Str message;
```

Comment:
[fix] Can we change this to `std::string` based on the char around that whole world, save revisiting own the not too distant.

Reply:
Updated in 0876e51a